File: sim_1_6_8.java
import java.io.File;
import java.io.PrintWriter;
import java.io.FileNotFoundException;
import java.util.Random;
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;
import org.apache.commons.cli.*;


class sim_1_6_8 {
    static Random rand = new Random();

    public static void main(String[] args) throws ParseException, FileNotFoundException {
        Options options = new Options();
        options.addOption("c", true, "Number of cases");
        options.addOption("d", true, "Maximum candidates");
        options.addOption("b", true, "Maximum ballots");
        options.addOption("o", true, "Output filename");


        CommandLineParser parser = new PosixParser();
        CommandLine cmd = parser.parse(options, args);
        int num_cases = 0;
        int max_candidates = 0;
        int max_ballots = 0;
        String output_file = null;
        if (cmd.hasOption("c"))
            num_cases = Integer.parseInt(cmd.getOptionValue("c"));
        if (cmd.hasOption("d"))
            max_candidates = Integer.parseInt(cmd.getOptionValue("d"));
        if (cmd.hasOption("b"))
            max_ballots = Integer.parseInt(cmd.getOptionValue("b"));
        if (cmd.hasOption("o"))
            output_file = cmd.getOptionValue("o");


        // print the number of cases
        PrintWriter writer = new PrintWriter(new File(output_file));
        String tmp = String.format("%d\n", num_cases);
        writer.write(tmp);
        System.out.print(tmp);


        String[] names =
        {
            "Zelda Roberto",
            "Pearle Bouffard",
            "Mahalia Pelkey",
            "Theodore Montijo",
            "Manual Hiebert",
            "Manuela Hoffmeister",
            "Loris Freund",
            "Annabelle Lasher",
            "Angele Mclemore",
            "Charlotte Dimas",
            "Graig Mancuso",
            "Frida Gatto",
            "Tamela Mcglone",
            "Latisha Freese",
            "Farrah Hemingway",
            "Hoa Messina",
            "Gussie Summa",
            "Alisia Vangorder",
            "Robyn Harbison",
            "Janelle Cusson"
        };

        for (int i = 0; i < num_cases; i++)
        {
            // print a blank line
            tmp = String.format("\n");
            writer.write(tmp);
            System.out.print(tmp);


            // print the number of candidates
            int num_candidates = rand.nextInt(max_candidates)+1;
            tmp = String.format("%d\n", num_candidates);
            writer.write(tmp);
            System.out.print(tmp);


            // print all the candidates' names, each on a line
            for (int j = 0; j < num_candidates; j++) {
                tmp = String.format("%s\n", names[j]);
                writer.write(tmp);
                System.out.print(tmp);
            }


            // print the ballots, each on a line
            BallotList ballots = new BallotList(num_candidates);
            int num_ballots = rand.nextInt(max_ballots) + 1;
            for (int j = 0; j < num_ballots; j++) {
                ballots.shuffle();
                ballots.print(writer);
            }
        }

        writer.close();
    }


    static class BallotList {
        final int MAX_BALLOTS = 400;
        List<Integer> _list;


        BallotList(int size) {
            _list = new ArrayList<Integer>();
            for (int i = 0; i < size; i++)
                _list.add(i+1);
        }


        void shuffle() {
/*
            // old way to shuffle
            if (_list.size() > 1) {
                for (int i = 0; i < _list.size()-1; i++) {
                    int j = rand.nextInt(_list.size());
                    int tmp = _list.get(j);
                    _list.set(j, _list.get(i));
                    _list.set(i, tmp);
                }
            }
*/
            // new way to shuffle
            Collections.shuffle(_list);
        }


        void print(PrintWriter writer) {
            for (int i = 0; i < _list.size(); i++) {
                String tmp = String.format("%d ", _list.get(i));
                writer.write(tmp);
                System.out.print(tmp);
            }
            writer.write("\n");
            System.out.print("\n");
        }
    }
}

File: sim_1_6_8.c
#include <stdio.h>
#include <time.h>
#include <unistd.h> /* optarg */
#include <stdlib.h> /* RAND_MAX */
#include <string.h> /* strcpy() */


#define MAX_CANDIDATES 20
#define MAX_NAME       80
#define MAX_BALLOTS    400


char names[MAX_CANDIDATES][MAX_NAME] =
{
    "Zelda Roberto",
    "Pearle Bouffard",
    "Mahalia Pelkey",
    "Theodore Montijo",
    "Manual Hiebert",
    "Manuela Hoffmeister",
    "Loris Freund",
    "Annabelle Lasher",
    "Angele Mclemore",
    "Charlotte Dimas",
    "Graig Mancuso",
    "Frida Gatto",
    "Tamela Mcglone",
    "Latisha Freese",
    "Farrah Hemingway",
    "Hoa Messina",
    "Gussie Summa",
    "Alisia Vangorder",
    "Robyn Harbison",
    "Janelle Cusson"
};


struct Ballot
{
    int _list[MAX_BALLOTS];
    int _count;
};
typedef struct Ballot ballot_t;


void b_init(ballot_t* ballot, int size)
{
    int i;
    for (i = 0; i < size; i++)
        ballot->_list[i] = i+1;
    ballot->_count = size;
}


void b_shuffle(ballot_t* ballot)
{
    if (ballot->_count > 1)
    {
        int i;
        for (i = 0; i < ballot->_count - 1; i++) 
        {
            int j = i + rand() / (RAND_MAX / (ballot->_count - i) + 1);
            int tmp = ballot->_list[j];
            ballot->_list[j] = ballot->_list[i];
            ballot->_list[i] = tmp;
        }
    }
}


void b_print(ballot_t* ballot, FILE* file)
{
    int i;
    for (i = 0; i < ballot->_count; i++)
        fprintf(file, "%d ", ballot->_list[i]);
    fprintf(file, "\n");
}


void main(int argc, char** argv)
{
    int num_cases = 0;
    int max_candidates = 0;
    int max_ballots = 0;
    char* output_file = NULL;
    char opt;
    while ((opt = getopt(argc, argv, "c:d:b:o:")) != -1)
    {
        switch (opt)
        {
        case 'c':
            num_cases = atoi(optarg);
            break;
        case 'd':
            max_candidates = atoi(optarg);
            break;
        case 'b':
            max_ballots = atoi(optarg);
            break;
        case 'o':
            output_file = optarg;
            break;
        default:
            printf("Usage: %s -c num_cases -d max_candidates -b max_ballots -o output_file\n", argv[0]);
        }
    }

    if (num_cases == 0 || output_file == NULL)
        printf("Usage: %s -c num_cases -d max_candidates -b max_ballots -o output_file\n", argv[0]);
    else
    {
        srand(time(NULL));

        FILE* file = fopen(output_file, "w");


        // print the number of cases
        fprintf(file, "%d\n", num_cases);
        int i;
        for (i = 0; i < num_cases; i++)
        {
            // print a blank line
            fprintf(file, "\n");


            // print the number of candidates
            int num_candidates = rand() % max_candidates + 1;
            fprintf(file, "%d\n", num_candidates);


            // print all the candidates' name, each name on one line
            int j;
            for (j = 0; j < num_candidates; j++)
                fprintf(file, "%s\n", names[j]);


            // print the ballots
            ballot_t ballots;
            b_init(&ballots, num_candidates);
            int num_ballots = rand() % max_ballots + 1;
            for (j = 0; j < num_ballots; j++)
            {
                b_shuffle(&ballots);
                b_print(&ballots, file);
            }
        }

        fclose(file);
    }
}
